cpp/src/arrow/compare.cc

```diff
@@ -53,6 +53,68 @@ using internal::checked_cast;
 
 namespace internal {
 
+// These helper functions assume we already checked the arrays have equal
+// sizes and null bitmaps.
+
+template <typename ArrowType, typename EqualityFunc>
+inline bool BaseFloatingEquals(const NumericArray<ArrowType>& left,
+                               const NumericArray<ArrowType>& right,
+                               EqualityFunc&& equals) {
+  using T = typename ArrowType::c_type;
+
+  const T* left_data = left.raw_values();
+  const T* right_data = right.raw_values();
+
+  if (left.null_count() > 0) {
+    for (int64_t i = 0; i < left.length(); ++i) {
+      if (left.IsNull(i)) continue;
+      // The comparison is written this way so that NaNs compare unequal
+      if (!equals(left_data[i], right_data[i])) {
+        return false;
+      }
+    }
+  } else {
+    for (int64_t i = 0; i < left.length(); ++i) {
+      if (!equals(left_data[i], right_data[i])) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+template <typename ArrowType>
+inline bool FloatingEquals(const NumericArray<ArrowType>& left,
+                           const NumericArray<ArrowType>& right, bool nans_equal) {
+  using T = typename ArrowType::c_type;
+
+  if (nans_equal) {
+    return BaseFloatingEquals<ArrowType>(left, right, [](T x, T y) -> bool {
+      return (x == y) || (std::isnan(x) && std::isnan(y));
+    });
+  } else {
+    return BaseFloatingEquals<ArrowType>(left, right,
+                                         [](T x, T y) -> bool { return x == y; });
+  }
+}
+
+template <typename ArrowType>
+inline bool FloatingApproxEquals(const NumericArray<ArrowType>& left,
+                                 const NumericArray<ArrowType>& right,
+                                 typename ArrowType::c_type epsilon, bool nans_equal) {
+  using T = typename ArrowType::c_type;
+
+  if (nans_equal) {
+    return BaseFloatingEquals<ArrowType>(
+        left, right, [epsilon](T x, T y) -> bool { return !(fabs(x - y) > epsilon); });
```

Comment:
Does this works with a NaN and a non-NaN? If not, it's worth adding a test case.

Reply:
Will do.

Reply:
This caught a logic error, thanks :-)

Reply:
Ternary logic is a pain (beware all the kernel we'll write!)